### DISCUSSION TITLE
android build configuration.

### DISCUSSION THREAD

**Discussion by balex499** [wix/Detox#4350] [Q&A]
Hello, community. Perhaps someone has encountered this problem. My application has a branched configuration with many productFlavors. I use the
` ./gradlew assembleUatRelease` 
command to build the main build.  (uat - one of flavor)
To get a test build I add 
`assembleAndroidTest -DtestBuildType=release` as well.
But in the end the builds have different signatures. And the emulator cannot run the application. 
**How can I explicitly configure AndroidTest build?**

**Comment by portyu9** [comment]
Hi @balex499 

With a product flavor such as `uat`, I would not use the generic:

```bash
./gradlew assembleAndroidTest
```

Build the exact application and test variants instead, for example:

```bash
./gradlew assembleUatRelease assembleUatReleaseAndroidTest -DtestBuildType=release
```

That follows the Detox pattern for flavored Android builds: the AndroidTest APK must correspond to the same flavor/build type as the app under test.

Also verify that Gradle is using the intended test build type, for example:

```gradle
android {
    testBuildType System.getProperty('testBuildType', 'debug')
}
```

Then point Detox's `binaryPath` at the actual `uatRelease` APK.

If Android still reports incompatible signatures, run Gradle's `signingReport` and compare the exact application and AndroidTest variants being generated. At that point the problem is not Detox choosing the wrong APK; the selected variants are actually being signed/configured differently.

For E2E, it can also be cleaner to create a dedicated release-like test build type instead of using the exact production signing configuration.

Please mark this answer as accepted if it helped, thank you.
